### PR TITLE
Add function calling SFT NeMo2.0 tutorial

### PR DESCRIPTION
**Pull request by Euynaheh**
# What does this PR do ?

Add function calling SFT NeMo2.0 tutorial.

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tutorials/llm/function_calling/nemo2-chat-sft-function-calling.ipynb [hunk 1]
```diff
@@ -124,6 +124,169 @@
     "\n",
     "* lr=1e-6 and 3 epoches.\n",
     "\n",
+    "Below is a data transformation script of [Salesforce/xlam-function-calling-60k](https://huggingface.co/datasets/Salesforce/xlam-function-calling-60k) for your reference."
```
Comment (jgerh) [inline tutorials/llm/function_calling/nemo2-chat-sft-function-calling.ipynb:127]: Here are the copyedits for the entire file with comments identified by line numbers (as inline commenting was not possible for most of the file). 

Line 9/9

add comma "LLM, while"

LLM-based agents can leverage the powerful comprehension and generation capabilities of LLMs, while also incorporating the planning and function-calling abilities to accomplish many complex tasks.

Line 11/11

change capitalization "Tool Learning" to lower case

In this tutorial, we will demonstrate how to perform SFT to learn function-calling (tool learning) using NeMo 2.0. 

Line 18/18

use imperative verb and capitalization for procedural headings

Step 1. Construct the Dataset

Line 43/43

fix word order "we should first"

To perform fine-tuning in NeMo 2.0, we should first transform the training dataset into a predefined format in NeMo 2.0. According to the different training strategies you use, there are two types of function-calling dataset formats:

Line 45/45

revise sentence, add definite articles (the), and fix punctuation

The first one is a single-turn function-calling dataset. For each piece of data, the assistant only calls the function once, and the conversation does not record the function-calling execution result. We focus on training the LLM for correct function selection and  correct function parameters generation.

Line 47/47

revise sentence, add definite articles (the), and fix punctuation

The second one is a multi-turn function-calling dataset. For each piece of data, the assistant calls the function once or more than once, and the conversation records the function-calling execution result. The assistant will make the next generation based on the function's return. It looks closer to a real-life scenario.

Line 49/49

add comma them, as

Both types of datasets work for function-calling SFT. You can choose one of them or a combination of them, as long as it's suitable for your target.

Line 51/51

use imperative verb and capitalization for procedural headings and add hyphens

Option 1: Use the Single-Turn Function-Calling Dataset

Line 53/53

remove indefinite article (a) and conjunction (as)

Let's take data from [glaiveai/glaive-function-calling-v2](https://huggingface.co/datasets/glaiveai/glaive-function-calling-v2) as an example. The original data format is shown below.

Line 72/72

add definite article (the) and revise

We should transform it into the NeMo chat dataset format, which consists of three fields: mask, system and conversations.

Line 74/74

fix capitalization

mask: The role that needs to be masked out to prevent the role from participating in loss calculation.

Line 76/76

fix capitalization

system: System prompt.

Line 78/78

fix capitalization

conversations: For each role, the conversation consists of two fields: "from" and "value."


Line 80/80

add definite article (the), add comma, and revise

We can transform the original data into the format shown below. Since we're constructing a single-turn function-calling dataset, you can end the conversation at the point of the tool call.

Line 117/117

revise for clarity, fix punctuation, NVIDIA is upper case

Using the single-turn function-calling datasets described above, we've successfully fine-tuned NVIDIA/Mistral-NeMo-Minitron-8B-Instruct based on NVIDIA/Mistral-NeMo-Minitron-8B-Base, which has a general function-calling ability. If you're interested, you can quickly experience its capabilities via NIM online. If you want to reproduce a model like NVIDIA/Mistral-NeMo-Minitron-8B-Instruct using NeMo, you can refer to the three open-source datasets we used. Note that we also used some internal datasets that are not open-sourced.


Line 127/290

use imperative verb and capitalization for procedural headings, add hyphens, and add definite article (the)

Option 2: Use the Multi-Turns Function-Calling Dataset


Line 129/292

fix punctuation (add hyphens) and add indefinite article (a)

For multi-turn function-calling dataset construction, the process is similar to constructing a single-turn function-calling dataset. The only difference is that we need to add one more role, 'Function,' to represent the function-calling return. Let's take the data below as an example.


Line 193/356

revise for clarity and add definite article (the)

We transform the original data into the NeMo chat format shown below. Note that, compared to the single-turn function-calling dataset, we add the 'Function' role in conversations to record the function-calling execution result. We should also mask out the user and function roles to prevent them from participating in loss calculation.

Line 242/405

use imperative verb for procedural headings

Step 2. Use NeMo-Run

Line 244/407

revise last line in paragraph for clarity

Despite the different dataset formats, whether it is a single-turn function-calling dataset or a multi-turn function-calling dataset, the training script using NeMo-Run remains the same.

Line 305/468

use imperative verb for procedural headings - suggested revision

Step 3. Evaluate the Trained Model


Line 307/470

revise for clarity, fix spelling, and change Hugging Face to two words

After successfully training a checkpoint, we should evaluate the effectiveness of the trained model. We can first convert the checkpoint back to a Hugging Face checkpoint, deploy inference, perform benchmarking, and verify the downstream tasks.


Line 309/472

use imperative verb for procedural headings - suggested revision

Deploy NeMo Inference


Line 311/474

"First" is preferred usage

First, we can quickly check the trained model's performance via NeMo inference.

Line 397/560

revise for clarity

If you're satisfied with the trained model's performance, we can continue. For the benchmark and downstream task assessment in the next two steps, the applications we will use only accept OpenAI API format inference requests. Therefore, we should first convert the saved checkpoint to a Hugging Face checkpoint for further deployment.

Line 422/585

sentence appears to be incomplete, see suggested revision to text

Then, we can follow the steps in TensorRT-LLM to establish an OpenAI API, allowing us to deploy the model and handle inference requests efficiently.

424/587

suggested revision to heading for clarity

Benchmark the Fine-Tuned LLM

Line 426/589

revise for clarity and fix punctuation (add hyphens)

To benchmark the function-calling ability of the fine-tuned LLM, you can refer to the Berkeley Function Call Leaderboard. You will get benchmark results similar to those shown below.

Line 449/612

use imperative verb for procedural headings - suggested revision

Perform Downstream Task Assessment

Line 451/614

revise second sentence

Since the returns of the function calls are hard-coded in this demo, we recommend you use the conversations below and input them in order:

Line 473/636

fix spelling and punctuation

Note that you should modify the LLM service configuration according to your model name and server address, such as http://10.123.123.123:8000/v1. If you want to try LLMs running on NIM online, you need to apply for a free API key and use the server address https://integrate.api.nvidia.com/v1.

**Review by cuichenx — APPROVED** [review]
